### PR TITLE
Feature/collapsible tree

### PR DESCRIPTION
**Pull request by Totoketchup**

### CODE REVIEW

**Comment by Totoketchup** [comment]
If you want to do a first review @Giska, it's ready. Or maybe you prefer to wait before adding this feature.

File: packages/react-craft-ai-decision-tree/src/components/edges.jsx [hunk 1]
```diff
@@ -1,31 +1,39 @@
...
+  position: relative;
   fill: none;
   stroke: ${SELECTED_COLOR_EDGES};
 `;
 
+
 class Edges extends React.Component {
+  constructor(props) {
+    super(props);
+  }
```
Comment (Giska) [inline packages/react-craft-ai-decision-tree/src/components/edges.jsx]: if props are not used to create state, not sure it is usefull

File: packages/react-craft-ai-decision-tree/src/components/edges.jsx [hunk 2]
```diff
@@ -64,58 +73,62 @@ class Edges extends React.Component {
...
+      .selectAll('path')
       .data(links, (d) => {
         d.linkClass = defaultLinksCssClass;
         if (edgePath.indexOf(d.target.id) !== -1) {
           d.linkClass = selectedLinksCssClass;
         }
         return d.target.id;
       });
-
+
```
Comment (Giska) [inline packages/react-craft-ai-decision-tree/src/components/edges.jsx]: trailing whitespace

File: packages/react-craft-ai-decision-tree/src/components/tree.jsx [hunk 3]
```diff
@@ -84,21 +84,72 @@ const TreeCanvas = styled('div')`
...
+    .nodeSize([NODE_WIDTH + NODE_WIDTH_MARGIN, NODE_HEIGHT]);
+  let nodes;
   let incr = 0;
 
-  // Compute the max tree depth(node which is the lowest leaf)
-  const enrichTreeRecursive = (index, node) => {
+  nodes = d3Hierarchy(root, (d) => d.children);
+
+  // Add the necessary information into the created d3Hierachy
+  const enrichHierarchyRecursive = (index, node) => {
```
Comment (Giska) [inline packages/react-craft-ai-decision-tree/src/components/tree.jsx]: Hierarchy is not very explicit for us, rename it enrichTreeRecursive

File: packages/react-craft-ai-decision-tree/src/components/edges.jsx [hunk 4]
```diff
@@ -64,39 +68,50 @@ class Edges extends React.Component {
     this.svgTreeRef = input;
   };
 
-  renderTree = (treeData, svgDomNode, nodes, links, edgePath, totalNbSamples, version, edgeType) => {
-    // Cleans up the SVG on re-render
+  renderTree = (treeData, nodes, links, edgePath, totalNbSamples, version, edgeType, width, height) => {
+    const svgDomNode = this.svgTreeRef;
     d3Select(svgDomNode)
```
Comment (Giska) [inline packages/react-craft-ai-decision-tree/src/components/edges.jsx]: I think you could use directly this.svgRefTree

File: packages/react-craft-ai-decision-tree/src/components/edges.jsx [hunk 5]
```diff
@@ -64,39 +68,50 @@ class Edges extends React.Component {
...
-
-    // Update the nodes…
-    let node = svg.selectAll('g.node')
+      .attr('width', width)
+      .attr('height', height);
+  
+    // ------------ NODES ------------
+
+    // Update the nodes
+    let nodesSvg = d3Select(svgDomNode)
```
Comment (Giska) [inline packages/react-craft-ai-decision-tree/src/components/edges.jsx]: same here

File: packages/react-craft-ai-decision-tree/src/components/edges.jsx [hunk 6]
```diff
@@ -64,39 +68,50 @@ class Edges extends React.Component {
...
+    
+    nodesSvg.exit()
       .remove();
 
-    // Update the links… (they are ordered)
-    let link = svg.selectAll('path.link')
+    // ------------ LINKS ------------
+
+    // Update the links
+    let link = d3Select(svgDomNode)
```
Comment (Giska) [inline packages/react-craft-ai-decision-tree/src/components/edges.jsx]: same here

File: packages/react-craft-ai-decision-tree/src/components/nodes.jsx [hunk 7]
```diff
@@ -23,18 +26,78 @@ const Links = styled('div')`
...
+  }
+  &:focus {
+    outline: none;
+  }
+  &:active {
+    background: ${SELECTED_COLOR_EDGES};
+  }
+`;
+
+
```
Comment (Giska) [inline packages/react-craft-ai-decision-tree/src/components/nodes.jsx]: remove empty line

File: packages/react-craft-ai-decision-tree/src/components/tree.jsx [hunk 8]
```diff
@@ -234,6 +295,41 @@ class Tree extends React.Component {
     }
   }
 
+  shouldComponentUpdate(nextProps, nextState) {
+    const doUpdate = !_.isEqual(nextProps, this.props)
```
Comment (Giska) [inline packages/react-craft-ai-decision-tree/src/components/tree.jsx]: return directly this test

File: packages/react-craft-ai-decision-tree/src/components/tree.jsx [hunk 9]
```diff
@@ -243,18 +339,19 @@ class Tree extends React.Component {
...
 
     // place correctly the tree in the svg with the minSvgWidth
-    _.forEach(nodes, (d) => {
+    _.forEach(nodeDescendantsArray, (d) => {
       d.x = d.x + offsetX;
       d.y = d.y + NODE_HEIGHT / 3; // take in account the height of the node above the link
     });
 
-    return { links, minSvgHeight, minSvgWidth, nodes, totalNbSamples };
+    return { links, minSvgHeight, minSvgWidth, nodeDescendantsArray, treed3, totalNbSamples };
```
Comment (Giska) [inline packages/react-craft-ai-decision-tree/src/components/tree.jsx]:
```
return {
...,
...
}
```

**Comment by Totoketchup** [comment]
@Giska linting done
@cloderic you told me you'd prefer to have the prediction in the non folded 'nodes' as well right ?
I can try to add some confidence formula as well.

**Comment by cloderic** [comment]
> @cloderic you told me you'd prefer to have the prediction in the non folded 'nodes' as well right ?
> I can try to add some confidence formula as well.

Yep that'd be nice, a first version w/o the confidence can work too.

**Comment by Totoketchup** [comment]
Screenshot 2019-04-08 at 10 51 51
Since I currently don't have confidence information, what background color do I put, grey ?

**Comment by Totoketchup** [comment]
Screenshot 2019-04-08 at 11 01 03
Here it is, @Giska @cloderic What do you think ?

**Comment by Giska** [comment]
Perhaps just the class ? To be like the leaf.

File: packages/react-craft-ai-decision-tree/src/components/nodeInformation/prediction.jsx [hunk 10]
```diff
@@ -37,13 +50,13 @@ const Prediction = ({ node, treeVersion, configuration }) => {
...
           <OutputDiv>
-            <code>{configuration.output[0]}</code> {value}
+            <code>{configuration.output[0]}</code> {outputValue}
           </OutputDiv>
-          <div>Confidence: {(confidence * 100).toFixed(2)}%</div>
-          {std ? <div>Standard deviation: {std.toFixed(2)}</div> : null}
+          {confidence ? <div>Confidence {(confidence * 100).toFixed(2)}%</div> : null}
+          {std ? <div>Standard deviation {std.toFixed(2)}</div> : null}
         </PredictionDiv>
       ) : null}
```
Comment (Giska) [inline packages/react-craft-ai-decision-tree/src/components/nodeInformation/prediction.jsx]:
```
{
  outputValue ? (
    <PredictionDiv>
      <OutputDiv>
        <code>{configuration.output[0]}</code> {outputValue}
      </OutputDiv>
      {
        confidence ? (
          <div>Confidence {(confidence * 100).toFixed(2)}%</div>
         ) : null
      }
      {
        std ? (
          <div>Standard deviation {std.toFixed(2)}</div>
        ) : null
      }
    </PredictionDiv>
  ) : null
}
```

File: packages/react-craft-ai-decision-tree/src/components/nodes.jsx [hunk 11]
```diff
@@ -50,41 +113,43 @@ class Nodes extends React.Component {
...
     const setSelectedNode = () => {
       if (this.props.updateSelectedNode) {
         this.props.updateSelectedNode(node.treePath);
       }
     };
 
+    const onClickExpand = () => {
+      if (!_.isNull(node.children)) {
+        node._children = node.children;
+        node.children = null;
```
Comment (Giska) [inline packages/react-craft-ai-decision-tree/src/components/nodes.jsx]: maybe a better name for node._children and node.children ?

File: packages/react-craft-ai-decision-tree/src/components/nodes.jsx [hunk 12]
```diff
@@ -120,36 +185,59 @@ class Nodes extends React.Component {
       this.setState({
         showingTooltip: true,
         tooltipText: text,
-        tooltipRef: this.nodeRef[index]
+        tooltipRef: !_.isUndefined(node.data.children) ? this.buttonRef[index] : this.nodeRef[index]
       });
     };
 
+    const PADDING = 15;
```
Comment (Giska) [inline packages/react-craft-ai-decision-tree/src/components/nodes.jsx]: This has nothing to do here

File: packages/react-craft-ai-decision-tree/src/components/nodes.jsx [hunk 13]
```diff
@@ -120,36 +185,59 @@ class Nodes extends React.Component {
...
-            (this.props.selectedNode === node.treePath ? SELECTED_BORDER_WIDTH : 0),
-          backgroundColor: color
-        }}
-      >
-        {text}
-      </Node>
+            top: -(this.props.selectedNode === node.treePath ? SELECTED_BORDER_WIDTH : 0),
+            left: -(this.props.selectedNode === node.treePath ? SELECTED_BORDER_WIDTH : 0),
+            backgroundColor: color
+          }}
```
Comment (Giska) [inline packages/react-craft-ai-decision-tree/src/components/nodes.jsx]: }} >

File: packages/react-craft-ai-decision-tree/src/components/nodes.jsx [hunk 14]
```diff
@@ -120,36 +185,59 @@ class Nodes extends React.Component {
...
+          padding: `${PADDING}px ${PADDING}px ${PADDING}px ${PADDING}px`,
+          position: 'absolute',
+          top:
+            node.y - NODE_HEIGHT / 3 - PADDING,
+          left:
+            node.x - NODE_WIDTH / 2 - PADDING
+        }}>
+        <Node
+          ref={ indexRef }
+          onMouseOver={ () => {
```
Comment (Giska) [inline packages/react-craft-ai-decision-tree/src/components/nodes.jsx]: no arrow function directly inside the component

File: packages/react-craft-ai-decision-tree/src/components/nodes.jsx [hunk 15]
```diff
@@ -120,36 +185,59 @@ class Nodes extends React.Component {
...
+            showTooltip();
+          } }
+          onMouseOut={ () => {
```
Comment (Giska) [inline packages/react-craft-ai-decision-tree/src/components/nodes.jsx]: same here

File: packages/react-craft-ai-decision-tree/src/components/tree.jsx [hunk 16]
```diff
@@ -111,29 +148,53 @@ function computeSvgSizeFromData(root) {
 
   nodes = enrichTreeRecursive(0, nodes);
 
-  minSvgHeight = (maxTreeDepth + 1) * NODE_DEPTH;
-  minSvgWidth = Math.abs(dxMin) + Math.abs(dxMax) + NODE_WIDTH;
+  if (!_.isUndefined(collapsedDepth)) {
+    // Collapse the node and all it's children
+    const collapse = (node) => {
```
Comment (Giska) [inline packages/react-craft-ai-decision-tree/src/components/tree.jsx]: can this function can be extract from the component ?

File: packages/react-craft-ai-decision-tree/src/components/tree.jsx [hunk 17]
```diff
@@ -111,29 +148,53 @@ function computeSvgSizeFromData(root) {
...
+    nodeDescendantsArray: nodeDescendantsArray,
+    links: nodes.links(),
+    treed3: nodes,
     totalNbSamples: nodeDescendantsArray[0].nbSamples,
-    offsetX: Math.abs(dxMin) + NODE_WIDTH / 2
+    offsetX: offsetX
   };
 }
 
+
```
Comment (Giska) [inline packages/react-craft-ai-decision-tree/src/components/tree.jsx]: remove

File: packages/react-craft-ai-decision-tree/src/components/tree.jsx [hunk 18]
```diff
@@ -196,6 +258,40 @@ class Tree extends React.Component {
...
+    this.setState(
+      {
+        nodeDescendantsArray: nodeDescendantsArray,
+        links: links,
+        minSvgHeight: minSvgHeight,
+        minSvgWidth: minSvgWidth,
+        treed3: treed3,
+        clickedNode: node
+      }
+    );
```
Comment (Giska) [inline packages/react-craft-ai-decision-tree/src/components/tree.jsx]:
```
this.setState({
  nodeDescendantsArray: nodeDescendantsArray,
  links: links,
  minSvgHeight: minSvgHeight,
  minSvgWidth: minSvgWidth,
  treed3: treed3,
  clickedNode: node
});
```

File: CHANGELOG.md [hunk 19]
```diff
@@ -7,6 +7,10 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 
 ## [Unreleased](https://github.com/craft-ai/react-craft-ai-components/compare/v0.0.22...HEAD)
 
+## Added
+
+- Trees are now foldable. The props `collapsedDepth` has been added to the `Tree` component setting the depth after which nodes are folded. A node can be expanded by clicking the `+` button under it and collapsed by clicking the `-` button.
+
```
Comment (Giska) [inline CHANGELOG.md]: you miss also the add of the display of the major class in the node

**Comment by Totoketchup** [comment]
`if possible the node when we click on the collapse button shouldn't move` this is super difficult I guess

**Comment by Totoketchup** [comment]
I change `node._children ` into `node.hidden_children `

**Pull request merged**